train_concept_identification.py

```diff
@@ -41,6 +42,91 @@ def compute_loss(criterion, logits, gold_outputs):
   return loss
 
+
+def compute_fScore(gold_outputs,
+                   predicted_outputs,
+                   vocabs: Vocabs):
+    """Computes f_score, precision, recall.
+
+  Args:
+    gold_outputs: Gold outputs, shape (output seq len, batch size)
+    predicted_outputs: Predicted outputs, shape (output seq len, batch size)
+    vocabs: Vocabs object
+  Returns:
+    f_score
+  """
+
+    eos_index= list(vocabs.concept_vocab.keys()).index(EOS)
+    concepts_as_list_predicted, concepts_as_list_gold = tensor_to_list(gold_outputs, predicted_outputs, eos_index, vocabs)
+
+    print("concepts_as_list_gold", concepts_as_list_gold)
+    print("concepts_as_list_predicted", concepts_as_list_predicted)
+
+    true_positive = len(set(concepts_as_list_gold) & set(concepts_as_list_predicted))
+    false_positive = len(set(concepts_as_list_predicted).difference(set(concepts_as_list_gold)))
+    false_negative = len(set(concepts_as_list_gold).difference(set(concepts_as_list_predicted)))
+    precision = 0
+    recall = 0
+    print(f'tp = {true_positive}, fn = {false_negative}, fp = {false_positive}')
+
+    if len(concepts_as_list_predicted) != 0:
+        precision = true_positive / (true_positive + false_positive)
+        recall = true_positive / (true_positive + false_negative)
+    f_score = 0
+
+    if precision + recall != 0:
+        f_score = 2 * (precision * recall) / (precision + recall)
+
+    return f_score
+
+
+def tensor_to_list(gold_outputs,
+                   predicted_outputs,
+                   eos_index,
+                   vocabs:Vocabs):
+    # Extract padding from original outputs
+    gold_list_no_padding = extract_padding(gold_outputs, eos_index)
+    predicted_list_no_padding = extract_padding(predicted_outputs, eos_index)
+
+    # Remove UNK from the sequence
+    concepts_as_list_gold = remove_unk_from_sequence(gold_list_no_padding, vocabs)
+    concepts_as_list_predicted = remove_unk_from_sequence(predicted_list_no_padding, vocabs)
+
+    return concepts_as_list_predicted, concepts_as_list_gold
+
+
+def extract_padding(outputs, eos_index):
+    list_with_padding = []
+    list_no_padding = []
+
+    # Transpose the tensors, transform them in lists and remove the root
+    for sentence in torch.transpose(outputs, 0, 1):
+        list_with_padding.append(sentence.tolist()[1:])
+
+    # Remove the padding -> stop at EOS, for both gold and predicted concepts
+    for sentence in list_with_padding:
+        sentence_no_padding = []
+        for word in sentence:
+            if int(word) == eos_index:
+                break
+            else:
+                sentence_no_padding.append(word)
+        list_no_padding.append(sentence_no_padding)
+    return list_no_padding
+
+
+def remove_unk_from_sequence(outputs_no_padding,
+                             vocabs: Vocabs):
+    # TODO store the gold data before numericalization and use it here
+    ids_to_concepts_list = list(vocabs.concept_vocab.keys())
```

Comment:
Either move this out of the function or rename the function
At this point the function converts indices to strings and removes UNK tokens
It's an unexpected behaviour to have a function called 'remove_unk_from_sequence' also do the denumericalization -- the other way around might be better
Either rename to denumericalize/indices_to_words/indinces_to_strings or smth along those lines or transform the indices to words before this call and only have this call remove the UNK tokens